src/Compilers/Core/Portable/InternalUtilities/EnumerableExtensions.cs

```diff
@@ -382,11 +382,30 @@ public static T[] AsArray<T>(this IEnumerable<T> source)
             return builder.ToImmutableAndFree();
         }
 
+        public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IEnumerable<TSource>? source, Func<TSource, IEnumerable<TResult>> selector)
+        {
+            if (source == null)
+                return ImmutableArray<TResult>.Empty;
+
+#if NET
+            var count = source.TryGetNonEnumeratedCount(out var val) ? val : 0;
+#else
+            var count = 0;
+#endif
+
+            var builder = ArrayBuilder<TResult>.GetInstance(count);
+            foreach (var item in source)
+                builder.AddRange(selector(item));
+
+            return builder.ToImmutableAndFree();
+        }
+
         public static ImmutableArray<TResult> SelectManyAsArray<TSource, TResult>(this IReadOnlyCollection<TSource>? source, Func<TSource, IEnumerable<TResult>> selector)
```

Comment:
existing method only worked on IReadOnlyCollection.  This meant we couldn't just call the helper on normal enumerables, or on things like dictionaries.

Reply:
Couldn't we check for IReadOnlyCollection in the method that takes IEnumerable (on desktop)?

Reply:
sure.  we could reimplement TryGetNonEnumeratedCount on desktop.  But i didn't feel like it was that valuable.  FWIW, the use of 'count' here is pretty suspect to begin with as the capacity/count of the final collection really is unrelated to the Count of hte original collection we're SelectMany'ing on.

Reply:
ok.  will remove.